Add support for standard webpush

p1gp1g opened the pull request:
Replacement for #18457 

This adds support for the standardized web push notifications __without any breaking change__. It allows applications supporting the standard to use it while keeping support for the legacy version.

This also fixes support for UnifiedPush, as it requires following webpush specifications. At this moment, many app have implemented workaround for this: they sync with the server when they receive a push notification, but this not really optimal, and may introduce bugs depending on how notifications are synced

I have also patched the webpush lib, it didn't follow the spec correctly.

The test uses the RFC example, to be sure everything works as expected

Documentation: https://github.com/mastodon/documentation/pull/1585

p1gp1g committed:
Add support for standard webpush

ThisIsMissEm commented on app/validators/web_push_key_validator.rb:
```diff
@@ -3,7 +3,7 @@
   def validate(subscription)
     begin
-      Webpush::Encryption.encrypt('validation_test', subscription.key_p256dh, subscription.key_auth)
+      LegacyWebpush::Encryption.encrypt('validation_test', subscription.key_p256dh, subscription.key_auth)
```
I'm not sure this needs to be changed, as really what we're wanting to assert here is that the supplied key is a valid p256dh key.

These two sections are nearly identical:
- [Standard](https://github.com/p1gp1g/webpush/blob/main/lib/webpush/encryption.rb#L8-L32)
- [Legacy](https://github.com/p1gp1g/webpush/blob/legacy/lib/legacy-webpush/encryption.rb#L5-L22)

So maybe we can replace that with checking this doesn't throw?

```
      group_name = "prime256v1"
      salt = Random.new.bytes(16)

      server = OpenSSL::PKey::EC.generate(group_name)
      server_public_key_bn = server.public_key.to_bn

      group = OpenSSL::PKey::EC::Group.new(group_name)
      client_public_key_bn = OpenSSL::BN.new(decode64(p256dh), 2)
      client_public_key = OpenSSL::PKey::EC::Point.new(group, client_public_key_bn)

      shared_secret = server.dh_compute_key(client_public_key)
```

Because I think that's where the exception is thrown, not when actually doing the encryption.

ThisIsMissEm replied:
(The only difference is the info tag)

p1gp1g replied:
When `LegacyWebpush` is used, it means I haven't change the classes, it uses the old one (that have been renamed)

This indeed just test if this is a valid P256 key. So we can use one or the other class

I can move to the new class (`Webpush`) and add a comment to explain that works for the legacy too

p1gp1g replied:
By the way, the error `crypto.errors.invalid_key` is wrong, it deals with ED25519/X25519 keys but that's not the case here. I will open an issue/pr for this. Since all the translations are wrong too, it should be moved to a new id isn't it ? Like `crypto.errors.invalid_p256_key` ?

ThisIsMissEm commented on db/schema.rb:
```diff
@@ -10,9 +10,9 @@
 #
```
This file shouldn't be changed

p1gp1g replied:
This is done automatically when running the migration. Because it requires adding the new `standard` field

p1gp1g replied:
Do you want I manually remove unnecessary changes (like the `t0.`)?

ClearlyClaire replied:
Yes please!

p1gp1g replied:
Done

ThisIsMissEm commented on lib/tasks/mastodon.rake:
```diff
@@ -45,7 +45,7 @@ namespace :mastodon do
       end
 
-      vapid_key = Webpush.generate_key
+      vapid_key = LegacyWebpush.generate_key
```
Does this method actually change between the two versions?

p1gp1g replied:
They are the same, and I kept the old one, but I can use the new one

ThisIsMissEm commented on app/workers/web/push_notification_worker.rb:
```diff
@@ -48,11 +60,38 @@ def perform(subscription_id, notification_id)
   end
 
-  private
+  def perform_standard_request
```
Rather than duplicating the entire request method, it'd be better to have like a `prepare_headers` and `prepare_payload` methods, leaving the rest the same.

p1gp1g commented:
@ThisIsMissEm Thank you for your review.

I've addressed most of your comments:
- I've changed the webpush tasks common to the legacy and standard to use the standard Webpush class
- I've move the duplicated code to a dedicated function

I've also push API changes I forgot to commit yesterday :+1:

ClearlyClaire reviewed (commented):
Hi, thank you for your contribution! This is much welcome, and while I have only skimmed through it yet, this seems like a sound approach.

One thing I'm slightly worried about here is the use of new dependencies, but I suppose that can't be avoided, unless we get the upstream `webpush` gem to expose methods to do both current and “legacy” webpush.

p1gp1g commented:
> One thing I'm slightly worried about here is the use of new dependencies, but I suppose that can't be avoided, unless we get the upstream `webpush` gem to expose methods to do both current and “legacy” webpush.

In order to avoid new project relying on the drafts of the specifications by mistake, I'd argue that it is a good thing that the upstream gem doesn't expose the legacy webpush :)

BTW, I 've [patched](https://github.com/p1gp1g/webpush/commit/124ec025eb42269c0c853a2fb92f5c8a480f6929) your version of the library, do you want I open a PR ?

ClearlyClaire commented:
I completely missed that upstream was unmaintained, so my talk of “unless we get the upstream `webpush` gem […]” is basically moot :sweat_smile:

I think I would still prefer a single gem to expose both the current-spec and “legacy” method we use (possibly named in a more explicit way) if that isn't too much work, but failing that I guess two branches is ok too.

> BTW, I 've [patched](https://github.com/p1gp1g/webpush/commit/124ec025eb42269c0c853a2fb92f5c8a480f6929) your version of the library, do you want I open a PR ?

Yes I think it would make sense to have that in my repo. Or better yet, have it be owned by the Mastodon project.

p1gp1g commented:
Any idea how to fix `PG::NotNullViolation: ERROR:  null value in column "standard" of relation "web_push_subscriptions" violates not-null constraint` even if I've set [`null: false, default: false`](https://github.com/mastodon/mastodon/pull/33528/files#diff-3cb00946053239107a9d4c93b2a7f79f523876c51bfa11c74ee6a4a32a5bf324R5) ?

p1gp1g commented:
> Ok, I've cloned the fork to https://github.com/mastodon/webpush so it is now maintained by Mastodon itself.

I've kept only openssl-3 and v0.3-openssl3, renamed "main" and "legacy" to help catch the versions. I can open PR for both

> Meanwhile, I'm re-implementing the old encryption as `Webpush::Legacy::Encryption` (suggestions for something else than `Legacy` to reference a particular webpush spec draft are welcome), this way we don't have to maintain two branches and we'll have less duplication.

Legacy is implementing 4th draft of RFC8291 + 1st draft of RFC8292, if it gives you some inspiration

ClearlyClaire commented on app/controllers/api/v1/push/subscriptions_controller.rb:
```diff
@@ -21,6 +21,7 @@ def create
         endpoint: subscription_params[:endpoint],
         key_p256dh: subscription_params[:keys][:p256dh],
         key_auth: subscription_params[:keys][:auth],
+        standard: subscription_params[:standard],
```
When omitted, `subscription_params[:standard]` is nil, which means you are building a hash with `standard: nil`. I would use something like `subscription_params[:standard] || false` or conditionally add it to the hash.

p1gp1g replied:
Thanks, will push a fix

ClearlyClaire commented on app/controllers/api/web/push_subscriptions_controller.rb:
```diff
@@ -76,6 +76,7 @@ def web_push_subscription_params
       endpoint: subscription_params[:endpoint],
       key_auth: subscription_params[:keys][:auth],
       key_p256dh: subscription_params[:keys][:p256dh],
+      standard: subscription_params[:standard],
```
Ditto!

ThisIsMissEm commented on app/models/web/push_subscription.rb:
```diff
@@ -5,10 +5,11 @@
 #  key_auth        :string           not null
-#  data            :json
+#  key_p256dh      :string           not null
+#  standard        :boolean          default(FALSE)
```
Another option here might be to use an enum with default:

```
enum :version, { legacy: 0, standard: 1 }
validates :version, inclusion: { in: versions.keys }
```

(I think I have that right)

p1gp1g replied:
I don't know if it helps a lot, since we know there won't be another version of the specifications ?

p1gp1g replied:
Regarding the default, it would be nice to define a date when the default will be true (can be in years), pushing for the adoption to the standard. What do you think ?

ThisIsMissEm commented on app/validators/web_push_key_validator.rb:
```diff
@@ -3,6 +3,8 @@
 class WebPushKeyValidator < ActiveModel::Validator
   def validate(subscription)
     begin
+      # We test if the subscription key is a valid p256 pubkey, we
+      # use Webpush but it would work with LegacyWebpush too (it also uses the same curve)
       Webpush::Encryption.encrypt('validation_test', subscription.key_p256dh, subscription.key_auth)
```
@p1gp1g / @ClearlyClaire perhaps we could add a `Webpush.keys_valid?(key_p256_dh, key_auth) #=> boolean` method? That way we're not actually messing about with the encryption here, but instead doing exactly what we intend.

The only reason I wrote this code this way was because it was the quickest way to ship _some_ validation, without reimplementing a bunch of stuff.

p1gp1g replied:
Trying to encrypt something looks like a good way to test the value to me :+1:

ClearlyClaire replied:
I'm not against a more specific test, but it does not need to be in this PR.

ClearlyClaire commented:
Looks like this is going right. Can you try with `gem 'webpush', github: 'mastodon/webpush', ref: 'f5f3281f99435a39ca3627c160bdaa9726c18b87'` instead of the two repos?

p1gp1g commented:
The legacy tests work but not standard one until https://github.com/ClearlyClaire/webpush/pull/1 is merged

ClearlyClaire commented:
Sorry, I thought it would result in easier-to-review changes. Let me know if you want me to deal with the conflicts.

ClearlyClaire commented:
That is fine to me, yes!

p1gp1g commented:
The PR is squashed and rebased

ClearlyClaire commented on app/serializers/rest/web_push_subscription_serializer.rb:
```diff
@@ -1,7 +1,11 @@
+  attributes :id, :endpoint, :standard, :alerts, :server_key, :policy
+
+  def standard
+    object.standard || false
```
I don't think `|| false` is needed here, as the column has `null: false`.

ClearlyClaire commented on config/environments/development.rb:
```diff
@@ -43,6 +43,7 @@
   end
 
   # Generate random VAPID keys
+  # Uses Webpush class, Webpush::Legacy does the same thing.
```
I think this comment can be removed, `Webpush::Legacy` (as in, in the gem maintained by Mastodon) does not cover this at all.

ClearlyClaire commented on config/environments/test.rb:
```diff
@@ -52,6 +52,7 @@
   config.x.otp_secret = '100c7faeef00caa29242f6b04156742bf76065771fd4117990c4282b8748ff3d99f8fdae97c982ab5bd2e6756a159121377cce4421f4a8ecd2d67bd7749a3fb4'
 
   # Generate random VAPID keys
+  # Uses Webpush class, Webpush::Legacy does the same thing.
```
I think this comment can be removed, `Webpush::Legacy` (as in, in the gem maintained by Mastodon) does not cover this at all.

ClearlyClaire commented on lib/tasks/mastodon.rake:
```diff
@@ -45,6 +45,7 @@ namespace :mastodon do
         env[key] = SecureRandom.alphanumeric(32)
       end
 
+      # Uses Webpush class, Webpush::Legacy does the same thing.
```
I think this can be removed, `Webpush::Legacy` (as in, in the gem maintained by Mastodon) does not cover this at all.

ClearlyClaire commented on lib/tasks/mastodon.rake:
```diff
@@ -571,6 +572,7 @@ namespace :mastodon do
   namespace :webpush do
     desc 'Generate VAPID key'
     task :generate_vapid_key do
+      # Uses Webpush class, Webpush::Legacy does the same thing.
```
I think this can be removed, `Webpush::Legacy` (as in, in the gem maintained by Mastodon) does not cover this at all.

ClearlyClaire reviewed (commented):
Thanks again for your contribution! This looks good overall!

I think the `standard` key to `subscription` is slightly awkward, especially since the value of `subscription` typically directly comes from browser APIs, but I don't really have a better suggestion, as the API endpoint is not versioned (should we maybe add a versioned API endpoint for this and deprecate the current one? But this probably should be a separate PR).

The code changes look good if slightly verbose because of the two code paths. I have verified (changing the service worker's code) that the new code path works with Firefox.

p1gp1g committed:
Remove unnecessary fallback

p1gp1g committed:
Remove unnecessary comments

p1gp1g committed:
Revert unnecessary changes

p1gp1g committed:
fix default value for standard

ClearlyClaire commented on db/schema.rb:
```diff
@@ -10,7 +10,7 @@
-ActiveRecord::Schema[7.2].define(version: 2024_12_16_224825) do
+ActiveRecord::Schema[8.0].define(version: 2025_01_08_111200) do
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
```
It looks like the change in your previous commit here was due to a Rails 8 change (we had no migration since our switch to Rails 8), though both this version and the previous one seem to work just fine.
```suggestion
  enable_extension "pg_catalog.plpgsql"
```

ClearlyClaire committed:
Apply suggestions from code review

Update generated annotations and schema.db

p1gp1g commented:
Thank you for the review and the merge ! This will be in 4.4.0 isn't it ?

ClearlyClaire commented:
Yes, it's not planned for the 4.3 branch, it'll be in the 4.4 one.

p1gp1g commented:
I'll open a PR for the webpush to FCM bridge after some tests

ShadowJonathan commented:
It is possible that notifications are still broken for Toot! and other apps, documented in https://github.com/glitch-soc/mastodon/issues/2981 for glitch-soc, but I'm still gathering data about that

ShadowJonathan commented:
Testing this on mastodon.social; Notifications are definitely not going through for `Toot!` and the official mastodon app.

p1gp1g commented:
@ShadowJonathan From the linked issue, it seems unrelated to this PR isn't it ?

ShadowJonathan commented:
@p1gp1g yes, apologies, we were investigating something, and at one point it appeared related to this issue, but ended up not

p1gp1g commented:
No problem, I just wanted to be sure and to clarify in case someone ends here for another issue related to webpush :)
